File: packages/ucm-core/src/markers/verifierResolver.ts
// Resolve the concrete verifier(s) a marker row's verification_policy demands.
//
// A verification_policy in "requirements" mode lists `required_verifiers` ids per
// requirement. Each id resolves to either (a) an explicit verifier declared under
// `policy.verifiers`, or (b) the default-convention "acceptance" verifier that
// runs the row's own use-case test. Anything else is returned BLOCKED rather than
// thrown, so callers can surface an actionable diagnostic instead of crashing.
//
// Pure + deterministic: given a row the result is sorted by verifier_id and
// deduped, and `{slug}` is substituted everywhere it appears in command/inputs.

export const DEFAULT_CONVENTION_VERIFIER_ID = "acceptance";

const SLUG_TOKEN = "{slug}";

// The default-convention "acceptance" verifier: run the row's own use-case test.
const DEFAULT_CONVENTION_COMMAND = [
  "pnpm",
  "-s",
  "vitest",
  "run",
  "tests/use-cases/{slug}.test.ts"
];
const DEFAULT_CONVENTION_INPUTS = ["tests/use-cases/{slug}.test.ts"];

// The minimal row shape this resolver needs: a slug (the row/use-case id, used to
// substitute {slug}) plus the raw verification_policy object.
export interface VerifierResolverRow {
  slug: string;
  verification_policy: unknown;
}

export interface ResolvedVerifier {
  verifier_id: string;
  status: "resolved";
  source: "policy" | "default_convention";
  kind: "script";
  evidence_kind: string;
  command: string[];
  inputs: string[];
  timeout_seconds?: number;
}

export interface BlockedVerifier {
  verifier_id: string;
  status: "blocked";
  reason: string;
}

export type VerifierResolution = ResolvedVerifier | BlockedVerifier;

// Resolve every required_verifier id referenced by the row's verification_policy.
// The returned array is deduped by verifier_id and sorted ascending for stability.
export function resolveRowVerifiers(row: VerifierResolverRow): VerifierResolution[] {
  const ids = collectRequiredVerifierIds(row.verification_policy);
  const verifiers = extractVerifiers(row.verification_policy);
  return ids
    .map((id) => resolveOne(id, row.slug, verifiers))
    .sort((left, right) =>
      left.verifier_id < right.verifier_id ? -1 : left.verifier_id > right.verifier_id ? 1 : 0
    );
}

function resolveOne(
  id: string,
  slug: string,
  verifiers: Record<string, unknown>
): VerifierResolution {
  const declared = verifiers[id];
  if (isRecord(declared)) {
    return resolveDeclared(id, slug, declared);
  }
  if (id === DEFAULT_CONVENTION_VERIFIER_ID) {
    return defaultConventionVerifier(id, slug);
  }
  return {
    verifier_id: id,
    status: "blocked",
    reason: `verifier '${id}' is not declared in verification_policy.verifiers and is not the default-convention id '${DEFAULT_CONVENTION_VERIFIER_ID}'`
  };
}

function resolveDeclared(
  id: string,
  slug: string,
  declared: Record<string, unknown>
): ResolvedVerifier {
  const resolved: ResolvedVerifier = {
    verifier_id: id,
    status: "resolved",
    source: "policy",
    kind: "script",
    evidence_kind:
      typeof declared.evidence_kind === "string" ? declared.evidence_kind : "test_result",
    command: toStringArray(declared.command).map((part) => substituteSlug(part, slug)),
    inputs: toStringArray(declared.inputs).map((part) => substituteSlug(part, slug))
  };
  if (typeof declared.timeout_seconds === "number") {
    resolved.timeout_seconds = declared.timeout_seconds;
  }
  return resolved;
}

function defaultConventionVerifier(id: string, slug: string): ResolvedVerifier {
  return {
    verifier_id: id,
    status: "resolved",
    source: "default_convention",
    kind: "script",
    evidence_kind: "test_result",
    command: DEFAULT_CONVENTION_COMMAND.map((part) => substituteSlug(part, slug)),
    inputs: DEFAULT_CONVENTION_INPUTS.map((part) => substituteSlug(part, slug))
  };
}

function collectRequiredVerifierIds(policy: unknown): string[] {
  if (!isRecord(policy) || policy.mode !== "requirements" || !Array.isArray(policy.requirements)) {
    return [];
  }
  const seen = new Set<string>();
  const ids: string[] = [];
  for (const requirement of policy.requirements) {
    if (!isRecord(requirement) || !Array.isArray(requirement.required_verifiers)) {
      continue;
    }
    for (const id of requirement.required_verifiers) {
      if (typeof id === "string" && !seen.has(id)) {
        seen.add(id);
        ids.push(id);
      }
    }
  }
  return ids;
}

function extractVerifiers(policy: unknown): Record<string, unknown> {
  if (isRecord(policy) && isRecord(policy.verifiers)) {
    return policy.verifiers;
  }
  return {};
}

function substituteSlug(value: string, slug: string): string {
  return value.split(SLUG_TOKEN).join(slug);
}

function toStringArray(value: unknown): string[] {
  return Array.isArray(value)
    ? value.filter((item): item is string => typeof item === "string")
    : [];
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

File: UseCasesCore/Sources/UseCasesCore/Markers/VerifierResolver.swift
/// Where a resolved verifier's entry was found.
public enum VerifierSource: String, Equatable, Sendable {
  case policy
  case workspaceConfiguration = "workspace_config"
  case workspaceDefault = "workspace_default"
}

/// A verifier the row's policy demands, resolved to a concrete script.
public struct ResolvedVerifier: Equatable, Sendable {
  public let verifierIdentifier: String
  public let source: VerifierSource
  public let evidenceKind: String
  public let command: [String]
  public let inputs: [String]
  public let timeoutSeconds: Double?
  /// The preset this verifier expanded from; nil for an explicit script. It is
  /// the only mechanical way to tell a named test runner from arbitrary argv.
  public let preset: VerifierPresetIdentifier?

  public init(
    verifierIdentifier: String,
    source: VerifierSource,
    evidenceKind: String,
    command: [String],
    inputs: [String],
    timeoutSeconds: Double?,
    preset: VerifierPresetIdentifier?,
  ) {
    self.verifierIdentifier = verifierIdentifier
    self.source = source
    self.evidenceKind = evidenceKind
    self.command = command
    self.inputs = inputs
    self.timeoutSeconds = timeoutSeconds
    self.preset = preset
  }

  /// The TypeScript object: `preset` is written before `timeout_seconds`,
  /// because the timeout is assigned after the literal is built.
  var jsonValue: JSONValue {
    var object = JSONObject([
      ("verifier_id", .string(verifierIdentifier)),
      ("status", .string("resolved")),
      ("source", .string(source.rawValue)),
      ("kind", .string("script")),
      ("evidence_kind", .string(evidenceKind)),
      ("command", .array(command.map(JSONValue.string))),
      ("inputs", .array(inputs.map(JSONValue.string))),
    ])
    object["preset"] = preset.map { preset in
      .string(preset.rawValue)
    }
    object["timeout_seconds"] = timeoutSeconds.map(JSONValue.number)
    return .object(object)
  }
}

/// A verifier that could not be resolved, with the actionable reason.
public struct BlockedVerifier: Equatable, Sendable {
  public let verifierIdentifier: String
  public let reason: String

  public init(
    verifierIdentifier: String,
    reason: String,
  ) {
    self.verifierIdentifier = verifierIdentifier
    self.reason = reason
  }

  var jsonValue: JSONValue {
    .object(JSONObject([
      ("verifier_id", .string(verifierIdentifier)),
      ("status", .string("blocked")),
      ("reason", .string(reason)),
    ]))
  }
}

public enum VerifierResolution: Equatable, Sendable {
  case resolved(ResolvedVerifier)
  case blocked(BlockedVerifier)

  public var verifierIdentifier: String {
    switch self {
    case let .resolved(verifier): verifier.verifierIdentifier
    case let .blocked(verifier): verifier.verifierIdentifier
    }
  }

  var jsonValue: JSONValue {
    switch self {
    case let .resolved(verifier): verifier.jsonValue
    case let .blocked(verifier): verifier.jsonValue
    }
  }
}

/// Resolve the concrete verifier(s) a row's verification_policy demands
/// (verifierResolver.ts). Each required id resolves, in order, from the row's
/// own `verifiers`, the workspace config's `verifiers`, and — for the default
/// convention id only — the entry the workspace's `default` names; otherwise it
/// is BLOCKED, returned rather than thrown.
public enum VerifierResolver {
  /// The id rows use for "the workspace default verifier". It has no built-in
  /// command.
  public static let defaultConventionVerifierIdentifier = "acceptance"

  /// Every required verifier id, deduplicated, resolved and sorted by id in
  /// UTF-16 code-unit order.
  public static func resolveRowVerifiers(
    slug: String,
    variant: String? = nil,
    verificationPolicy: JSONValue?,
    workspace: ResolvedWorkspaceVerifiers = ResolvedWorkspaceVerifiers(),
  ) -> [VerifierResolution] {
    let rowVerifiers = verificationPolicy?["verifiers"]?.objectValue ?? JSONObject()
    let context = ResolutionContext(
      slug: slug,
      variant: variant,
      rowVerifiers: rowVerifiers,
      workspace: workspace,
    )
    let resolutions = requiredVerifierIdentifiers(verificationPolicy).map(context.resolve)
    return sortedByIdentifier(resolutions)
  }

  /// Stable, by `verifier_id` in code-unit order.
  static func sortedByIdentifier(_ resolutions: [VerifierResolution]) -> [VerifierResolution] {
    resolutions.enumerated()
      .sorted { left, right in
        let leftIdentifier = left.element.verifierIdentifier
        let rightIdentifier = right.element.verifierIdentifier
        if JavaScriptString.precedes(leftIdentifier, rightIdentifier) {
          return true
        }
        if JavaScriptString.precedes(rightIdentifier, leftIdentifier) {
          return false
        }
        return left.offset < right.offset
      }
      .map(\.element)
  }

  private static func requiredVerifierIdentifiers(_ policy: JSONValue?) -> [String] {
    guard let policy = policy?.objectValue,
          policy["mode"] == .string("requirements"),
          let requirements = policy["requirements"]?.arrayValue
    else {
      return []
    }
    var seen = OrderedStringSet()
    for requirement in requirements {
      guard let identifiers = requirement["required_verifiers"]?.arrayValue else {
        continue
      }
      for identifier in identifiers.compactMap(\.stringValue) {
        seen.insert(identifier)
      }
    }
    return seen.members
  }
}

private struct ResolutionContext {
  let slug: String
  let variant: String?
  let rowVerifiers: JSONObject
  let workspace: ResolvedWorkspaceVerifiers

  func resolve(_ identifier: String) -> VerifierResolution {
    if let entry = rowVerifiers[identifier]?.objectValue {
      return resolveEntry(identifier, entry, source: .policy)
    }
    if let entry = workspaceEntry(identifier) {
      return resolveEntry(identifier, entry, source: .workspaceConfiguration)
    }
    if JavaScriptString.identical(identifier, VerifierResolver.defaultConventionVerifierIdentifier),
       let defaultIdentifier = workspace.defaultVerifierIdentifier
    {
      if let target = workspaceEntry(defaultIdentifier) {
        return resolveEntry(identifier, target, source: .workspaceDefault)
      }
      return .blocked(BlockedVerifier(
        verifierIdentifier: identifier,
        reason: "workspace verifiers.default '\(defaultIdentifier)' "
          + "does not name a declared verifier",
      ))
    }
    return .blocked(BlockedVerifier(
      verifierIdentifier: identifier,
      reason: "no verifier '\(identifier)' configured; declare it in the row's "
        + "verification_policy.verifiers or the workspace verifiers map, or set verifiers.default",
    ))
  }

  /// The workspace entry whose id has exactly these code units. Not a
  /// dictionary subscript: Swift's `String` keys match canonically equivalent
  /// ids, which the TypeScript's property lookup never does.
  private func workspaceEntry(_ identifier: String) -> JSONObject? {
    workspace.verifiers.first { pair in
      JavaScriptString.identical(pair.key, identifier)
    }?.value.value
  }

  private func resolveEntry(
    _ identifier: String,
    _ entry: JSONObject,
    source: VerifierSource,
  ) -> VerifierResolution {
    let evidenceKind = entry["evidence_kind"]?.stringValue ?? "test_result"
    let timeoutSeconds = entry["timeout_seconds"]?.numberValue
    guard let presetIdentifier = entry["preset"]?.stringValue else {
      return .resolved(ResolvedVerifier(
        verifierIdentifier: identifier,
        source: source,
        evidenceKind: evidenceKind,
        command: substituted(entry["command"]),
        inputs: substituted(entry["inputs"]),
        timeoutSeconds: timeoutSeconds,
        preset: nil,
      ))
    }
    switch VerifierPresets
      .expand(presetIdentifier: presetIdentifier, slug: slug, variant: variant)
    {
    case let .blocked(reason):
      return .blocked(BlockedVerifier(verifierIdentifier: identifier, reason: reason))
    case let .resolved(preset, expansion):
      let inputs = entry["inputs"]?.arrayValue == nil
        ? expansion.inputs
        : substituted(entry["inputs"])
      return .resolved(ResolvedVerifier(
        verifierIdentifier: identifier,
        source: source,
        evidenceKind: evidenceKind,
        command: expansion.command,
        inputs: inputs,
        timeoutSeconds: timeoutSeconds,
        preset: preset,
      ))
    }
  }

  /// The string members of an array, tokens substituted; `[]` for anything
  /// that is not an array.
  private func substituted(_ value: JSONValue?) -> [String] {
    (value?.arrayValue ?? []).compactMap(\.stringValue).map { part in
      VerifierPresets.substituteTokens(part, slug: slug, variant: variant)
    }
  }
}
